Subsection on how to obtain more than just the posterior from Turing.jl (#89)
arviz-devs/ArviZ.jl: docs/src/quickstart.md

    τ ~ truncated(Cauchy(0, 5), 0, Inf)
    θ = TV(undef, J)
    θ .~ Normal(μ, τ)

>     for i in eachindex(y)

Author: Nah, not faster. But for `elementwise_loglikelihood` it's not clear what `MvNormal` in this case means. Is `y` a single observation or is it a vector of observations? And so if you want to use `elementwise_loglikelihood`, you need to be "explicit" about which you mean in the model definition, unfortunately :confused:

>     for i in eachindex(y)

Reviewer: Ah, my question was more about `for i in eachindex(y)` vs `for i in 1:J`, since `J` is a passed parameter that is assumed to be the length of `y` and `σ`. IIRC at some time in the past, the advice in Turing was to explicitly have size parameters like `J` when possible because of some efficiency boost.

>     for i in eachindex(y)

Reviewer: That makes sense though about `elementwise_loglikelihood`.

>     for i in eachindex(y)

Author: Ah, uhh, I don't think so? The only reason why I could see that being the case is due to constant propagation? Other than that, I'm not sure if there's going to be much of a difference. I think you might know more about this than me, but could it make a difference for AD if the number of iterations for the for-loop is known at compile time?

>     for i in eachindex(y)

Reviewer: Yeah, I don't think so. If `J` is a passed parameter, I don't think its value would be known at compile time anyways. And in fact if `y` is statically sized, then `eachindex(y)` maybe could cause the loop to be unrolled. I don't think either case necessarily is better for AD.

>     for i in eachindex(y)

Author: > If J is a passed parameter, I don't think its value would be known at compile time anyways.

The evaluator for the model is a closure created when you instantiate the model, so at least in theory `J` is known at compile time (whether or not it is indeed treated as a static variable by the Julia compiler I'm not sure, though I would expect so).

> And in fact if y is statically sized,

Yeah I was thinking that too.

gcf()
```

### Additional information in Turing.jl

> In the above `from_mcmcchains` call, we did not specify the `prior`, `posterior_predictive`, nor the `log_likelihoods`. This section describes how one would go about obtaining these from the `Model` and `Chains`, which we can then pass to `from_mcmcchains` to construct the [`InferenceData`](@ref).

Author: Much better :+1:

To sample from the prior, one simply calls `sample` but with the `Prior` sampler:
```@example quickstart
prior = sample(param_mod, Prior(), nsamples; progress = false)
```

> Obtaining the predictive posterior is a matter of first instantiating a "predictive model", i.e. a `Model` but with the observations set to `missing`, and then call `predict` on the predictive model and the inferred posterior:

Reviewer: We can use the same approach to draw from the prior predictive distribution, right?
Also, could we be explicit that the user isn't getting the distribution but a sample from the distribution?
Maybe something like
```suggestion
To draw from the prior and posterior predictive distributions we can instantiate a "predictive model", i.e. a Turing model but with the observations set to `missing`, and then calling `predict` on the predictive model and the previously drawn samples:
```

Author: Yep! And I prefer your suggestion, so let's go with that:)

```@example quickstart
# Instantiate the predictive model
param_mod_predict = turing_model(
    J, fill(missing, length(y)), σ

> )

Author: Yeah ofc, that's much nicer! Added.

# and then sample!

> posterior_predictive = predict(param_mod_predict, turing_chns)

Reviewer: This should work, right?
```suggestion
prior_predictive = predict(param_mod_predict, prior)
posterior_predictive = predict(param_mod_predict, turing_chns)
```

Author: Forgot about the prior predictive; added!

```
And to extract the elementwise log-likelihoods, which is useful if you want to compute metrics such as [`loo`](@ref),
```@example quickstart
loglikelihoods = elementwise_loglikelihoods(param_mod, turing_chns)
```
This can then be included in the `from_mcmcchains` call from above:
```@example quickstart
# Convert into format compatible with ArviZ.jl
using LinearAlgebra

> loglikelihoods_arr = permutedims(cat(values(loglikelihoods)...; dims = 3), (2, 1, 3))

Author: Ah, yeah that's a good point! I didn't think about the ordering as I at the time I was mainly interested in `loo`. For the tutorial, maybe we should just match the keys of `loglikelihoods` with the predictive posterior `Chains`, e.g.
```julia
loglikelihoods_vals = getindex.(Ref(loglikelihoods), string.(keys(posterior_predictive)))
loglikelihoods_arr = permutedims(cat(loglikelihoods_vals...; dims = 3), (2, 1, 3))
```

> I'm curious, why does Turing return a Dict here instead of a Chains?

A couple of reasons. 
1. `elementwise_loglikelihood` is part of `DynamicPPL` which doesn't use MCMCChains.jl, hence we wanted to use a structure which was independent of MCMCChains.jl. But of course we could just wrap this in Turing.jl, which would then return a `Chains`, which brings me to the second point.
2. I'd argue that the elementwise loglikelihoods doesn't quite fit into a `Chains` structure, _as it is now_. We could of course just add fields `y[1]_loglikelihood` or something, but that will make things cluttered quickly + we'll have to implement something hacky so the user can access to the elementwise loglikelihoods directly from the chain (e.g. `elementwise_loglikelihoods(c::Chains)` which finds the corresponding fields using `Regex`). The _correct_ approach I think would be something similar to `InferenceData` where the loglikelihoods is a "special" field in the chain. But at the moment, this sort of goes against the approach taken in MCMCChains.jl where no fields are treated in a special manner; it's just a matter of whether or not it's a `parameter` or `internal`. And if we made this an `internal` field, we would still have to wrap it in some other container to map observation variable name to the loglikelihood.

Thinking about it again, maaaybe it's not such a bad idea to return a chain where the only fields present are the loglikelihoods (I was initially thinking of returning a chain from `predict` with both the sampled values and the loglikelihoods), e.g. `y[1]` now represents the loglikelihoods. I'm not a huge fan though as it could be confusing to look at a chain with the keys being the same as in the chain obtained from `sample` but now representing something different. So I think I'm leaning towards something along the lines as I mentioned in (2) above.

> loglikelihoods_arr = permutedims(cat(values(loglikelihoods)...; dims = 3), (2, 1, 3))

Reviewer: Thanks! That helps a lot. For me, it would be handy of log-likelihoods were not returned in a generic data structure like a `Dict`, because then I can't infer anything about them from the type, hence the need below to manually reshape, which is annoying and could be error-prone for the user.

> Thinking about it again, maaaybe it's not such a bad idea to return a chain where the only fields present are the loglikelihoods (I was initially thinking of returning a chain from predict with both the sampled values and the loglikelihoods), e.g. y[1] now represents the loglikelihoods. I'm not a huge fan though as it could be confusing to look at a chain with the keys being the same as in the chain obtained from sample but now representing something different. So I think I'm leaning towards something along the lines as I mentioned in (2) above.

Here are my general thoughts. First, I don't think it needs to be confusing to have log-likelihoods in a data structure meant for random samples. After all, because the log-likelihoods are deterministically computed from random samples and constants, the resulting log-likelihoods are in fact random samples. The big downside I see is, yeah, I think it could be confusing for the log-likelihoods to have the same name as the variable, but since the user is explicitly calling an `elementwise_loglikelihoods` function, and the values aren't just being injected into another function call, that doesn't seem like a big deal if documented. It's just that the user really should never merge such a `Chains` object with a posterior or prior sample. I'd have to think more about the (2) approach.

> loglikelihoods_arr = permutedims(cat(values(loglikelihoods)...; dims = 3), (2, 1, 3))

Author: >  because then I can't infer anything about them from the type

So the return-value from `elementwise_loglikelihood` is always going to be the same type, i.e. `Dict{String, Matrix{Float64}}`, so the processing that we do in this readme could be automated. I get what you're saying though, and I agree.

> First, I don't think it needs to be confusing to have log-likelihoods in a data structure meant for random samples. 

Yeah I agree with this. My issue is that I don't think the _currrent_ `MCMCChains.Chains` is suitable for also having the `loglikelihoods` together with the samples :confused: 

>  It's just that the user really should never merge such a Chains object with a posterior or prior sample. I'd have to think more about the (2) approach.

The thing is I'm worried that this could happen at some point though, e.g. it seems natural to allow returning the loglikelihoods from `predict` in addition to the predicted "observations".

idata = from_mcmcchains(
    turing_chns,

>     prior = prior,
>     posterior_predictive = posterior_predictive,
>     coords = Dict("school" => schools),
>     dims = Dict(
>         "y" => ["school"],
>         "σ" => ["school"],
>         "θ" => ["school"],
>     ),
>     library = "Turing",
>     log_likelihood = loglikelihoods_arr

Reviewer: Minor point, but can we match the ordering of the groups in the returned object?
```suggestion
    posterior_predictive = posterior_predictive,
    log_likelihood = loglikelihoods_arr,
    prior = prior,
    prior_predictive = prior_predictive,
    coords = Dict("school" => schools),
    dims = Dict(
        "y" => ["school"],
        "σ" => ["school"],
        "θ" => ["school"],
    ),
    library = "Turing",```

Author: Done!
(Sorry about not just straight up committing these changes, thus resolving these comments. I just wanted to run everything locally to make sure it worked, and then I added the code-changes in one go.)

Reviewer: No problem at all! Thanks!

)
```

> 

Reviewer: Can you just give one sentence introducing LOO?
Also, it would be nice to show a plot after this, maybe
```julia
plot_loo_pit(idata; y = "y", ecdf = true)
```

Author: Good idea; I added a couple of sentences + plot

gcf()
```

### Additional information in Turing.jl

> With a few more steps, we can use Turing to compute additional useful groups to add to the [`InferenceData`].(@ref).

Reviewer: ```suggestion
With a few more steps, we can use Turing to compute additional useful groups to add to the [`InferenceData`](@ref).
```

To sample from the prior, one simply calls `sample` but with the `Prior` sampler:
```@example quickstart
prior = sample(param_mod, Prior(), nsamples; progress = false)
```
To draw from the prior and posterior predictive distributions we can instantiate a "predictive model", i.e. a Turing model but with the observations set to `missing`, and then calling `predict` on the predictive model and the previously drawn samples:
```@example quickstart
# Instantiate the predictive model
param_mod_predict = turing_model(J, similar(y, Missing), σ)
# and then sample!
prior_predictive = predict(param_mod_predict, prior)
posterior_predictive = predict(param_mod_predict, turing_chns)
```
And to extract the elementwise log-likelihoods, which is useful if you want to compute metrics such as [`loo`](@ref),
```@example quickstart
loglikelihoods = Turing.elementwise_loglikelihoods(param_mod, turing_chns)
```

> This can then be included in the `from_mcmcchains` call from above:

Reviewer: Might be nice to provide the link here:
```suggestion
This can then be included in the [`from_mcmcchains`](@ref) call from above:
```

```@example quickstart

> # Convert into format compatible with ArviZ.jl

Reviewer: It's a bit confusing as to whether this comment applies to the entire block or just the conversion of log-likelihoods. Could you clarify?

Author: Maybe we could just remove this line now, since we added the two comments below explaining what's going on?

Reviewer: Yeah that works

Author: ```suggestion
```

using LinearAlgebra
# Ensure the ordering of the loglikelihoods matches the ordering of `posterior_predictive`

> loglikelihoods_vals = getindex.(Ref(loglikelihoods), string.(keys(posterior_predictive)))

Reviewer: Would this be more digestible (I don't know about the variable name `ynames`, naming things is hard).
```suggestion
ynames = string.(keys(posterior_predictive))
loglikelihoods_vals = getindex.(Ref(loglikelihoods), ynames)
```

> # Reshape into `(num_chains, num_samples, num_params)`

Reviewer: Would it be helpful here to use the variable names defined in this page?
```suggestion
# Reshape into `(nchains, nsamples, size(y)...)`
```

Author: That's up to you :+1: I think I maybe prefer `num_params` since it indicates what to do in general, while the code tells you how this particular example looks.

Also, it should probably be
```suggestion
# Reshape into `(num_chains, num_samples, length(y))`
```
instead?

Reviewer: Well if we were going with completely general, ArviZ uses `(nchain, ndraw, shape...)`. We here `size` instead of `shape` (because `np.shape` is the same as `size`). So the version I gave above is closer to the general version, and I'd prefer it. 

But I also should go through and replace `nchains` with `nchain` and `nsamples` with `ndraw` in a later PR.

Author: Ah, aight. I'll let you choose this one :+1:  
Hmm, yeah so it's actually slightly annoying to recover the correct shape from `MCMCChains.Chains` AFAIK. That's why I just went with the elementwise one.

Reviewer: Yeah, makes sense, the [code we have](https://github.com/arviz-devs/ArviZ.jl/blob/master/src/mcmcchains.jl#L32-L40) for inferring size from `Chains` relies on the indices and supports the indexing conventions of CmdStan and Turing (last I checked CmdStan uses dots instead of brackets to indicate indexing), but it could probably be improved. I'd rather not make that functionality part of the API, so I think you did the right thing here.

What we could always do is add a `from_turing` method consistent with the other `from_XXX` methods that forwards everything to `from_mcmcchains` but first pre-processes `log_likelihood`. That's the cleanest solution I can think of right now.

loglikelihoods_arr = permutedims(cat(loglikelihoods_vals...; dims = 3), (2, 1, 3))

idata = from_mcmcchains(
    turing_chns,
    posterior_predictive = posterior_predictive,

>     log_likelihood = loglikelihoods_arr,

Reviewer: I missed it earlier, but when passing an array to `InferenceData`, we need to tell it if it which variable in `observed_data` it should map to:
```suggestion
    log_likelihood = Dict("y" => loglikelihoods_arr),
```

    prior = prior,
    prior_predictive = prior_predictive,
    observed_data = Dict("y" => y),
    coords = Dict("school" => schools),
    dims = Dict(
        "y" => ["school"],
        "σ" => ["school"],
        "θ" => ["school"],
    ),
    library = "Turing"
)
```

Then we can for example compute the expected *leave-one-out (LOO)* predictive density, which is an estimate of the out-of-distribution predictive fit of the model:

```@example quickstart
loo(idata) # higher is better
```

> If the model is well-calibrated, i.e. it replicates the true generative process well, the CDF of the pointwise LOO values should be similarly distributed to a `Uniform` distribution. This can be inspected visually:

Reviewer: By Blue Style, we have exactly one sentence per line in markdown files (better for version control).
Also, should we avoid typing `Uniform` here since we're referring to a uniform distribution in the abstract, not a specific implementation?
```suggestion
If the model is well-calibrated, i.e. it replicates the true generative process well, the CDF of the pointwise LOO values should be similarly distributed to a uniform distribution.
This can be inspected visually:
```

```@example quickstart

> plot_loo_pit(idata; y = "y", ecdf = true)

Reviewer: An unfortunate quirk of using PyPlot.jl is that in many contexts, to show the plot one needs to call `gcf()` (get current figure). Hence we do that visibly here:
```suggestion
plot_loo_pit(idata; y = "y", ecdf = true);
gcf()
```
